### PR TITLE
clusteroperator: add controller from capi

### PR DESCRIPTION
**Pull request by zirain**
Signed-off-by: hejianpeng <hejianpeng2@huawei.com>

**What type of PR is this?**

/kind feature

**What this PR does / why we need it**:

1. import controllers form capi
1. upgrade capa to v2
1. improve feature gates

**Which issue(s) this PR fixes**:
Fixes #

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:

```release-note

```

### CODE REVIEW

File: Makefile [hunk 1]
```diff
@@ -46,7 +46,7 @@ build: clean tidy kurator cluster-operator
 
 .PHONY: tidy
 tidy:
-	go mod tidy -compat=1.17
+	go mod tidy -go=1.19
```
Comment (hzxuzhonghu) [inline Makefile:49]: why do we need it?

File: cmd/cluster-operator/capi/capi.go [hunk 2]
```diff
@@ -0,0 +1,243 @@
...
+	}
+
+	return nil
+}
+
+func setupReconcilers(ctx context.Context, opts *config.Options, mgr ctrl.Manager) error {
+	// Set up a ClusterCacheTracker and ClusterCacheReconciler to provide to controllers
+	// requiring a connection to a remote cluster
+	log := ctrl.Log.WithName("remote").WithName("ClusterCacheTracker")
+	tracker, err := remote.NewClusterCacheTracker(
```
Comment (hzxuzhonghu) [inline cmd/cluster-operator/capi/capi.go:62]: what does tracker do

File: cmd/cluster-operator/capi/capi.go [hunk 3]
```diff
@@ -0,0 +1,243 @@
...
+			Log:     &log,
+			Indexes: remote.DefaultIndexes,
+		},
+	)
+	if err != nil {
+		return fmt.Errorf("unable to create cluster cache tracker, %w", err)
+	}
+	if err := (&remote.ClusterCacheReconciler{
+		Client:           mgr.GetClient(),
+		Log:              ctrl.Log.WithName("remote").WithName("ClusterCacheReconciler"),
```
Comment (hzxuzhonghu) [inline cmd/cluster-operator/capi/capi.go:74]: same here, can you comment on what does these  controller actually do, so could know why they are needed

File: cmd/cluster-operator/scheme/aws.go [hunk 4]
```diff
@@ -19,36 +19,29 @@ package scheme
 
 import (
 	cgscheme "k8s.io/client-go/kubernetes/scheme"
-	infrav1alpha3 "sigs.k8s.io/cluster-api-provider-aws/api/v1alpha3"
-	infrav1alpha4 "sigs.k8s.io/cluster-api-provider-aws/api/v1alpha4"
-	infrav1 "sigs.k8s.io/cluster-api-provider-aws/api/v1beta1"
-	eksbootstrapv1alpha3 "sigs.k8s.io/cluster-api-provider-aws/bootstrap/eks/api/v1alpha3"
-	eksbootstrapv1alpha4 "sigs.k8s.io/cluster-api-provider-aws/bootstrap/eks/api/v1alpha4"
```
Comment (hzxuzhonghu) [inline cmd/cluster-operator/scheme/aws.go:26]: Assume this is removing the conversion between versions, if so do you also remove the that in CRD?

Comment (zirain) [reply cmd/cluster-operator/scheme/aws.go:26]: `awsProviderVersion := env("AWS_PROVIDER_VERSION", "v2.0.0")`

File: Makefile (hunk 1 above)
Comment (zirain) [reply Makefile:49]: update upgrade capa to v2.0.0, we can remove this.

File: cmd/cluster-operator/capi/capi.go (hunk 2 above)
Comment (zirain) [reply cmd/cluster-operator/capi/capi.go:63]: `ClusterCacheTracker.GetClient return client for remote cluster`

Comment (hzxuzhonghu) [reply cmd/cluster-operator/capi/capi.go:63]: I mean what is the responsibility of the controller

File: cmd/cluster-operator/scheme/aws.go (hunk 4 above)
Comment (hzxuzhonghu) [reply cmd/cluster-operator/scheme/aws.go:26]: ?

Comment (zirain) [reply cmd/cluster-operator/scheme/aws.go:26]: after capa upgraded to v2.0.0, all crds are re-generated.

File: cmd/cluster-operator/capi/capi.go (hunk 2 above)
Comment (zirain) [reply cmd/cluster-operator/capi/capi.go:63]: update comment

File: cmd/cluster-operator/Dockerfile [hunk 5]
```diff
@@ -2,6 +2,9 @@ ARG BASE_VERSION
 ARG BASE_IMAGE
 FROM ${BASE_IMAGE}:${BASE_VERSION}
 
-COPY cluster-operator /usr/local/bin/
-RUN chmod 750 /usr/local/bin/cluster-operator
-ENTRYPOINT ["/usr/local/bin/cluster-operator"]
+# Use uid of nonroot user (65532) because kubernetes expects numeric user when applying pod security policies
+USER 65532
```
Comment (hzxuzhonghu) [inline cmd/cluster-operator/Dockerfile:6]: what does this mean

File: cmd/cluster-operator/README.md [hunk 6]
```diff
@@ -27,13 +27,19 @@ helm install -n cert-manager cert-manager jetstack/cert-manager --set installCRD
 kubectl create namespace kurator-system
 helm install kurator-base out/charts/base-0.1.0.tgz -n kurator-system
 helm install kurator-cluster-operator out/charts/cluster-operator-0.1.0.tgz -n kurator-system
+helm install -f /root/capa-values.yaml -n kurator-system kurator-cluster-operator out/charts/cluster-operator-0.1.0.tgz
```
Comment (hzxuzhonghu) [inline cmd/cluster-operator/README.md:30]: The docs should go to docs dir

**Review by hzxuzhonghu — COMMENTED** [review]
Mostly lgtm

File: cmd/cluster-operator/README.md (hunk 6 above)
Comment (zirain) [reply cmd/cluster-operator/README.md:30]: this is a doc for dev, will add doc for cluster operator later.

File: cmd/cluster-operator/Dockerfile (hunk 5 above)
Comment (zirain) [reply cmd/cluster-operator/Dockerfile:6]: sync with upstream

**Review by hzxuzhonghu — APPROVED** [review]
/lgtm

**Pull request merged**